src/RecurringPurchase.sol

```diff
@@ -0,0 +1,270 @@
+ * @author Legend Labs, Inc.
+ */
+contract RecurringPurchase is QuarkScript {
+    using SafeERC20 for IERC20;
+
+    error BadPrice();
+    error InvalidInput();
+    error PurchaseConditionNotMet();
+
+    /// @notice Emitted when a swap is executed
+    event SwapExecuted(
+        address indexed sender,
+        address indexed recipient,
+        address tokenIn,
+        address tokenOut,
+        uint256 amountIn,
+        uint256 amountOut,
+        bytes path
+    );
+
+    /// @notice The base slippage factor where `1e18` represents 100% slippage tolerance
+    uint256 public constant BASE_SLIPPAGE_FACTOR = 1e18;
+
+    /**
+     * @dev Note: This script uses the following storage layout in the QuarkStateManager:
+     *         mapping(bytes32 hashedPurchaseConfig => uint256 nextPurchaseTime)
+     *             where hashedPurchaseConfig = keccak256(PurchaseConfig)
+     */
+
+    /// @notice Parameters for a recurring purchase order
+    struct PurchaseConfig {
+        uint256 interval;
+        SwapParams swapParams;
+        SlippageParams slippageParams;
+    }
+
+    /// @notice Parameters for performing a swap
+    struct SwapParams {
+        address uniswapRouter;
+        address recipient;
+        address tokenIn;
+        address tokenOut;
+        /// @dev The amount for tokenIn if exact in; the amount for tokenOut if exact out
+        uint256 amount;
+        /// @dev False for exact in; true for exact out
+        bool isExactOut;
+        uint256 deadline;
+        bytes path;
+    }
+
+    /// @notice Parameters for controlling slippage in a swap operation
+    struct SlippageParams {
+        /// @dev Maximum acceptable slippage, expressed as a percentage where 100% = 1e18
+        uint256 maxSlippage;
+        /// @dev Price feed addresses for determining market exchange rates between token pairs
+        /// Example: For SUSHI -> SNX swap, use [SUSHI/ETH feed, SNX/ETH feed]
+        address[] priceFeeds;
+        /// @dev Flags indicating whether each corresponding price feed should be inverted
+        /// Example: For USDC -> ETH swap, use [true] with [ETH/USD feed] to get ETH per USDC
+        bool[] shouldReverses;
+    }
+
+    /// @notice Cancel the recurring purchase for the current nonce
+    function cancel() external {
+        // Not explicitly clearing the nonce just cancels the replayable txn
+    }
+
+    /**
+     * @notice Execute a swap given a configuration for a recurring purchase
+     * @param config The configuration for a recurring purchase order
+     */
+    function purchase(PurchaseConfig calldata config) public {
+        allowReplay();
+
+        if (config.slippageParams.priceFeeds.length == 0) {
+            revert InvalidInput();
+        }
+        if (config.slippageParams.priceFeeds.length != config.slippageParams.shouldReverses.length) {
+            revert InvalidInput();
+        }
+
+        bytes32 hashedConfig = _hashConfig(config);
```

Comment:
Looks like here stores a timestamp with hashedConfig as a key to handle time interval, and the config has below unique fields.
```
config.swapParams.uniswapRouter,
config.swapParams.recipient,
config.swapParams.tokenIn,
config.swapParams.tokenOut,
config.swapParams.amount,
config.swapParams.isExactOut,
config.swapParams.deadline,
config.swapParams.path
```

If user added `buy 1ETH every day` then 1 day later, user created another recurring buy request and it's also `buy 1 ETH every day`

Would those configs conflict?  (those two seems will result in exact same set of parameters. Deadline are probably is uint256.max all the time?)  In this case, I think expect behavior will be like two same recurring buy running in parallel.


Probably, one way we could solve this is we "maintain" the `deadline` as always to be unique to the time of the first submission. Or add a additional field that's unique on the time of first submission probably. 🤔

Reply:
That new recurring purchase request will be a different Quark Operation with its own nonce right? If so, there will be no storage conflict because the storage in `StateManager` is isolated and nonce-scoped.

Reply:
oh I see.
So in that case we probably don't need to hash a key to read but even some constant string value key would work too?